File: php/src/ErrorTrackerClient.php
<?php

declare(strict_types=1);

namespace Royaltics\ErrorTracker;

use Royaltics\ErrorTracker\Core\EventBuilder;
use Royaltics\ErrorTracker\Core\Transport;
use Royaltics\ErrorTracker\Types\ClientConfig;
use Royaltics\ErrorTracker\Types\EventIssue;
use Royaltics\ErrorTracker\Types\EventLevel;
use Royaltics\ErrorTracker\Utils\Compression;

final class ErrorTrackerClient
{
    private readonly EventBuilder $eventBuilder;
    private readonly Transport $transport;
    private array $eventQueue = [];
    private bool $isActive = false;
    private bool $isEnabled;
    private bool $isProcessing = false;

    public function __construct(private readonly ClientConfig $config)
    {
        $this->isEnabled = $this->config->enabled;
        
        $this->eventBuilder = new EventBuilder(
            app: $this->config->app,
            version: $this->config->version,
            platform: $this->config->platform,
            device: $this->config->licenseDevice
        );

        $this->transport = new Transport($this->config);
    }

    public function start(): self
    {
        if ($this->isActive) {
            return $this;
        }

        $this->attachErrorHandlers();
        $this->startBatchProcessor();
        $this->isActive = true;

        return $this;
    }

    public function error(
        \Throwable $error,
        EventLevel $level = EventLevel::ERROR,
        ?array $metadata = null
    ): self {
        if (!$this->isEnabled) {
            return $this;
        }

        try {
            $title = $error->getMessage() ?: 'Unknown error';
            $event = $this->eventBuilder->build($title, $error, $level, $metadata);
            $this->enqueue($event);
        } catch (\Exception $e) {
            $this->handleInternalError('Failed to track error', $e);
        }

        return $this;
    }

    public function event(
        string $title,
        EventLevel $level = EventLevel::INFO,
        ?array $metadata = null
    ): self {
        if (!$this->isEnabled) {
            return $this;
        }

        try {
            $error = new \Exception($title);
            $event = $this->eventBuilder->build($title, $error, $level, $metadata);
            $this->enqueue($event);
        } catch (\Exception $e) {
            $this->handleInternalError('Failed to track event', $e);
        }

        return $this;
    }

    public function forceFlush(): void
    {
        while (count($this->eventQueue) > 0) {
            $this->processBatch();
        }
    }

    public function pause(): self
    {
        $this->isEnabled = false;
        return $this;
    }

    public function resume(): self
    {
        $this->isEnabled = true;
        return $this;
    }

    public function shutdown(): void
    {
        $this->isEnabled = false;
        $this->isActive = false;
        $this->forceFlush();
    }

    private function attachErrorHandlers(): void
    {
        set_error_handler(function (int $errno, string $errstr, string $errfile, int $errline): bool {
            $this->error(
                new \ErrorException($errstr, 0, $errno, $errfile, $errline),
                EventLevel::ERROR,
                ['source' => 'error_handler']
            );
            return false;
        });

        set_exception_handler(function (\Throwable $exception): void {
            $this->error($exception, EventLevel::FATAL, ['source' => 'exception_handler']);
            $this->forceFlush();
        });

        register_shutdown_function(function (): void {
            $error = error_get_last();
            if ($error !== null && in_array($error['type'], [E_ERROR, E_PARSE, E_CORE_ERROR, E_COMPILE_ERROR])) {
                $this->error(
                    new \ErrorException($error['message'], 0, $error['type'], $error['file'], $error['line']),
                    EventLevel::FATAL,
                    ['source' => 'shutdown_handler']
                );
                $this->forceFlush();
            }
        });
    }

    private function startBatchProcessor(): void
    {
        register_tick_function(function (): void {
            if (!$this->isActive) {
                return;
            }

            static $lastFlush = 0;
            $now = (int)(microtime(true) * 1000);

            if ($now - $lastFlush >= $this->config->flushInterval) {
                $this->processBatch();
                $lastFlush = $now;
            }
        });
    }

    private function enqueue(EventIssue $event): void
    {
        $this->eventQueue[] = $event;

        if (count($this->eventQueue) >= $this->config->maxQueueSize) {
            $this->processBatch();
        }
    }

    private function processBatch(): void
    {
        if (count($this->eventQueue) === 0 || $this->isProcessing) {
            return;
        }

        $this->isProcessing = true;

        try {
            $batch = array_splice($this->eventQueue, 0, $this->config->maxQueueSize);

            foreach ($batch as $event) {
                try {
                    $this->dispatchEvent($event);
                } catch (\Exception $e) {
                    $this->handleInternalError('Failed to dispatch event', $e);
                }
            }
        } finally {
            $this->isProcessing = false;
        }
    }

    private function dispatchEvent(EventIssue $event): void
    {
        $eventString = $this->eventBuilder->stringify($event);
        $compressed = Compression::compressAndEncode($eventString);
        $this->transport->send($compressed);
    }

    private function handleInternalError(string $context, \Exception $error): void
    {
        error_log("[ErrorTracker] {$context}: {$error->getMessage()}");
    }
}

File: csharp/ErrorTrackerClient.cs
using System.Collections.Concurrent;
using Royaltics.ErrorTracker.Core;
using Royaltics.ErrorTracker.Types;
using Royaltics.ErrorTracker.Utils;

namespace Royaltics.ErrorTracker;

public sealed class ErrorTrackerClient : IDisposable
{
    private readonly ClientConfig _config;
    private readonly EventBuilder _eventBuilder;
    private readonly Transport _transport;
    private readonly ConcurrentQueue<EventIssue> _eventQueue;
    private readonly CancellationTokenSource _cancellationTokenSource;
    private readonly SemaphoreSlim _processingLock;
    private readonly ConcurrentDictionary<string, long> _fingerprintCache;
    
    private bool _isActive;
    private bool _isEnabled;
    private bool _isProcessing;
    private Task? _flushTask;

    public ErrorTrackerClient(ClientConfig config)
    {
        config.Validate();
        
        _config = config;
        _isEnabled = config.Enabled;
        _eventQueue = new ConcurrentQueue<EventIssue>();
        _cancellationTokenSource = new CancellationTokenSource();
        _processingLock = new SemaphoreSlim(1, 1);
        _fingerprintCache = new ConcurrentDictionary<string, long>();

        _eventBuilder = new EventBuilder(
            config.App,
            config.Version,
            config.Platform,
            config.LicenseDevice
        );

        _transport = new Transport(config);
    }

    public ErrorTrackerClient Start()
    {
        if (_isActive)
            return this;

        AttachErrorHandlers();
        StartQueueProcessor();
        _isActive = true;

        return this;
    }

    public ErrorTrackerClient Error(
        Exception error,
        EventLevel level = EventLevel.ERROR,
        Dictionary<string, string>? metadata = null)
    {
        if (!_isEnabled)
            return this;

        try
        {
            var title = error.Message ?? "Unknown error";
            var eventIssue = _eventBuilder.Build(title, error, level, metadata);
            
            if (IsDuplicate(eventIssue))
                return this;

            Enqueue(eventIssue);
        }
        catch (Exception ex)
        {
            HandleInternalError("Failed to track error", ex);
        }

        return this;
    }

    public ErrorTrackerClient Event(
        string title,
        EventLevel level = EventLevel.INFO,
        Dictionary<string, string>? metadata = null)
    {
        if (!_isEnabled)
            return this;

        try
        {
            var error = new Exception(title);
            var eventIssue = _eventBuilder.Build(title, error, level, metadata);

            if (IsDuplicate(eventIssue))
                return this;

            Enqueue(eventIssue);
        }
        catch (Exception ex)
        {
            HandleInternalError("Failed to track event", ex);
        }

        return this;
    }

    private bool IsDuplicate(EventIssue eventIssue)
    {
        if (!_config.Deduplicate) return false;

        var fingerprint = GenerateFingerprint(eventIssue);
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        if (_fingerprintCache.TryGetValue(fingerprint, out var lastSeen) && 
            (now - lastSeen) < _config.DeduplicationInterval)
        {
            return true;
        }

        _fingerprintCache[fingerprint] = now;

        // Optional: cleanup old entries
        if (_fingerprintCache.Count > 1000)
        {
            foreach (var key in _fingerprintCache.Keys)
            {
                if (_fingerprintCache.TryGetValue(key, out var time) && now - time > _config.DeduplicationInterval)
                {
                    _fingerprintCache.TryRemove(key, out _);
                }
            }
        }

        return false;
    }

    private string GenerateFingerprint(EventIssue eventIssue)
    {
        var culprit = eventIssue.Context.Culprit ?? "unknown";
        var message = eventIssue.Event.Message ?? "";
        return $"{eventIssue.Title}:{message}:{culprit}";
    }

    public async Task ForceFlushAsync()
    {
        while (!_eventQueue.IsEmpty)
        {
            await ProcessQueueAsync();
        }
    }

    public ErrorTrackerClient Pause()
    {
        _isEnabled = false;
        return this;
    }

    public ErrorTrackerClient Resume()
    {
        _isEnabled = true;
        return this;
    }

    public void Shutdown()
    {
        _isEnabled = false;
        _isActive = false;

        _cancellationTokenSource.Cancel();
        _flushTask?.Wait();

        ForceFlushAsync().Wait();
    }

    public void Dispose()
    {
        Shutdown();
        _transport?.Dispose();
        _cancellationTokenSource?.Dispose();
        _processingLock?.Dispose();
    }

    private void AttachErrorHandlers()
    {
        AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
        {
            if (args.ExceptionObject is Exception ex)
            {
                Error(ex, EventLevel.FATAL, new Dictionary<string, string>
                {
                    ["source"] = "UnhandledException"
                });
                ForceFlushAsync().Wait();
            }
        };

        TaskScheduler.UnobservedTaskException += (sender, args) =>
        {
            Error(args.Exception, EventLevel.ERROR, new Dictionary<string, string>
            {
                ["source"] = "UnobservedTaskException"
            });
            args.SetObserved();
        };
    }

    private void StartQueueProcessor()
    {
        _flushTask = Task.Run(async () =>
        {
            while (!_cancellationTokenSource.Token.IsCancellationRequested)
            {
                try
                {
                    await Task.Delay(_config.FlushInterval, _cancellationTokenSource.Token);
                    await ProcessQueueAsync();
                }
                catch (OperationCanceledException)
                {
                    break;
                }
                catch (Exception ex)
                {
                    HandleInternalError("Queue processing failed", ex);
                }
            }
        }, _cancellationTokenSource.Token);
    }

    private void Enqueue(EventIssue eventIssue)
    {
        _eventQueue.Enqueue(eventIssue);

        if (_eventQueue.Count >= _config.MaxQueueSize)
        {
            _ = Task.Run(ProcessQueueAsync);
        }
    }

    private async Task ProcessQueueAsync()
    {
        if (_eventQueue.IsEmpty || _isProcessing)
            return;

        await _processingLock.WaitAsync();

        try
        {
            _isProcessing = true;

            while (_eventQueue.TryDequeue(out var eventIssue))
            {
                var success = false;
                Exception? lastError = null;

                // Try up to MaxRetries + 1 (initial + retries)
                for (var attempt = 0; attempt <= _config.MaxRetries; attempt++)
                {
                    try
                    {
                        await DispatchEventAsync(eventIssue);
                        success = true;
                        break;
                    }
                    catch (Exception ex)
                    {
                        lastError = ex;
                        if (attempt < _config.MaxRetries)
                        {
                            await Task.Delay(_config.ThrottleInterval, _cancellationTokenSource.Token);
                        }
                    }
                }

                if (!success)
                {
                    Console.Error.WriteLine($"[ErrorTracker] Failed to dispatch event after {_config.MaxRetries + 1} attempts. Event: \"{eventIssue.Title}\". Reason: {lastError?.Message}. This event will be ignored to prevent queue congestion.");
                }

                if (!_eventQueue.IsEmpty)
                {
                    await Task.Delay(_config.ThrottleInterval, _cancellationTokenSource.Token);
                }
            }
        }
        finally
        {
            _isProcessing = false;
            _processingLock.Release();
        }
    }

    private async Task DispatchEventAsync(EventIssue eventIssue)
    {
        var eventString = _eventBuilder.Stringify(eventIssue);
        var compressed = Compression.CompressAndEncode(eventString);
        await _transport.SendAsync(compressed, _cancellationTokenSource.Token);
    }

    private static void HandleInternalError(string context, Exception error)
    {
        Console.Error.WriteLine($"[ErrorTracker] {context}: {error.Message}");
    }
}
